### PR TITLE
Indicator refactor

### PR DESCRIPTION
**Pull request by ColinEberhardt**
Details can be found here: #378

### CODE REVIEW

**Comment by chrisprice** [comment]
Some miscellaneous, I'm looking at it on my phone so haven't been able to try it out yet:

# The structure of the computers is a little weird, they look configurable but aren't rebound.
# The name computers doesn't suggest the functionality to me. 
# Should the renderers allow a root accessor to be specified in case you've stashed the result in a nested object? It seems a likely use-case for multi-valued indicators.

**Comment by ColinEberhardt** [comment]
Wow - big text!

> The structure of the computers is a little weird, they look configurable but aren't rebound.

Yeah, I haven't looked at the component API yet, I just wanted to give people a chance to see what it looks like from an end-user perspective when these components are split out.

> The name computers doesn't suggest the functionality to me

Totally agree - I've been struggling with that one. Any ideas?

> Should the renderers allow a root accessor to be specified in case you've stashed the result in a nested object? It seems a likely use-case for multi-valued indicators.

I'm not sure I follow you there! It's also worth noting that some indicators, those that output a single series, would not have dedicated renderers, you'd just plug the output straight into a series:

```
var ma = fc.indicators.computers.movingAverage();
ma(data); 

var line = fc.series.line()
        .xScale(dateScale)
        .yScale(priceScale);

container.append('g')
        .datum(data)
        .call(line);
```

**Comment by chrisprice** [comment]
Opps thought that would give a numbered list... Apparently not!

Transformers, mutators, combinators... They still don't sound right though.

I'll have a go at creating an example tomorrow. But, in the last example I
don't understand how the line series knows the y value?

**Comment by ColinEberhardt** [comment]
Probably a mistake in the example ;-)

On Tue, May 26, 2015 at 9:04 pm, ScottLogic/d3-financial-components
<reply@reply.github.com>
wrote:
Opps thought that would give a numbered list... Apparently not!

Transformers, mutators, combinators... They still don't sound right though.

I'll have a go at creating an example tomorrow. But, in the last example I
don't understand how the line series knows the y value?

On Tue, 26 May 2015 20:36 Colin Eberhardt <notifications@github.com> wrote:

> Wow - big text!
>
> The structure of the computers is a little weird, they look configurable
> but aren't rebound.
>
> Yeah, I haven't looked at the component API yet, I just wanted to give
> people a chance to see what it looks like from an end-user perspective when
> these components are split out.
>
> The name computers doesn't suggest the functionality to me
>
> Totally agree - I've been struggling with that one. Any ideas?
>
> Should the renderers allow a root accessor to be specified in case you've
> stashed the result in a nested object? It seems a likely use-case for
> multi-valued indicators.
>
> I'm not sure I follow you there! It's also worth noting that some
> indicators, those that output a single series, would not have dedicated
> renderers, you'd just plug the output straight into a series:
>
> var ma = fc.indicators.computers.movingAverage();
> ma(data);
>
> var line = fc.series.line()
> .xScale(dateScale)
> .yScale(priceScale);
>
> container.append('g')
> .datum(data)
> .call(line);
>
> —
> Reply to this email directly or view it on GitHub
>
<https://github.com/ScottLogic/d3-financial-components/pull/382#issuecomment-105643561>
> .
>
—
Reply to this email directly or view it on GitHub
[https://github.com/ScottLogic/d3-financial-components/pull/382#issuecomment-105650096] .[https://github.com/notifications/beacon/ABDBft4Dtewld-zudt6uVq_R2ZuTsiWKks5oNMk1gaJpZM4EqZcH.gif]

**Comment by ColinEberhardt** [comment]
I've rolled-out the pattern to all of the indicators, and re-bound properties where appropriate.

**Comment by ColinEberhardt** [comment]
- [ ] add series to a `renderers` namespace, computers to a `algorithms` namespace and algorithms to `algorithms.calculators` namescape.
 - [ ] By default, merge calculator output to a single property, rather than flattening
 - [ ] algorithms should return the merged output, for more flexibility
 - [ ] add design guidelines

**Comment by ColinEberhardt** [comment]
@chrisprice - I think that covers everything - I'm happy with the result :-)

File: src/indicators/design-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+The following is a brief set of design guidelines for indicators:
+
+## Calculators
+
+ - calculators should operate on an input array of data, transforming it into a new array containing the indicator output values.
```
Comment (jleft) [inline src/indicators/design-guidelines.md]: Very pedantic, but some of the list items start with an uppercase letter and some with a lowercase letter. Doesn't really matter much though! :-)

File: src/indicators/design-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ - calculators should expose a `value` accessor property (or a number of accessors if required), which is used to extract values from the source array. For calculators that only depend on a single value for their input this should be the identity function.
+ - calculators should not have to handle undefined values in the input array of data.
+ - the length of the output array should be the same as the input array.
+ - if a calculator creates 'undefined' values, the value used should be configurable.
```
Comment (jleft) [inline src/indicators/design-guidelines.md]: If _undefined_ is referring to the JavaScript type, is it worth surrounding it with backticks, to show it refers to code?

Comment (ColinEberhardt) [inline src/indicators/design-guidelines.md]: No, it refers to anything that represents the concept of being undefined. See bolinger for example:

https://github.com/ScottLogic/d3-financial-components/blob/master/src/indicators/algorithms/bollingerBands.js#L9

File: src/indicators/design-guidelines.md (hunk 1 above)
Comment (ColinEberhardt) [inline src/indicators/design-guidelines.md]: Thanks for the review, I'll fix the casing!

File: src/indicators/algorithms/bollingerBands.js [hunk 3]
```diff
@@ -3,38 +3,21 @@
...
-                var avg = d3.mean(values);
-                var stdDev = d3.deviation(values);
-                var multiplierValue = multiplier.apply(this, arguments);
-                return {
-                    upper: avg + multiplierValue * stdDev,
-                    average: avg,
-                    lower: avg - multiplierValue * stdDev
-                };
-            });
+        var mergeFunction = function(datum, boll) { datum.bollingerBands = boll; };
```
Comment (chrisprice) [inline src/indicators/algorithms/bollingerBands.js]: Nitpic: inline?

**Pull request merged**